File: typescript/nth-prime/nth-prime.ts
function* possiblePrimes() {
    let n = 6

    for (; ;) {
        yield n - 1
        yield n + 1

        n += 6
    }
}

function* primes() {
    yield 2
    yield 3

    for (const prime of possiblePrimes()) {
        if (isPrime(prime)) {
            yield prime
        }
    }
}

function isPrime(n: number) {
    const r = Math.floor(Math.sqrt(n))

    if (r < 5) {
        return true
    }

    for (let i = 5; i < r + 1; i += 1) {
        if (n % i === 0) {
            return false
        }
    }

    return true
}

export default class Prime {
    public nth(n: number) {
        if (n < 1) {
            throw new Error('Prime is not possible')
        }

        let prime = 0
        const p = primes()

        for (let i = 0; i < n; i += 1) {
            prime = p.next().value
        }

        return prime
    }
}

File: nim/nth-prime/nth_prime.nim
import math, sequtils

proc isPrime(n: int): bool =
  let r = n.toFloat.sqrt.floor.toInt
  (5 .. r).allIt(n mod it != 0)

iterator primes: int =
  yield 2
  yield 3

  for r in countup(6, int.high, 6):
    if isPrime(r - 1):
      yield r - 1
    
    if isPrime(r + 1):
      yield r + 1

proc prime*(n: int): int =
  if n < 1:
    raise newException(ValueError, "Cannot use zero or negative number")
  
  var nth = 0

  for prime in primes():
    nth.inc

    if nth == n:
      return prime
